### ISSUE TITLE
Progress of A3C branch

### ISSUE THREAD

**Issue by deontologician** [keras-rl/keras-rl#21]
Is this experimental because it doesn't work, or because it hasn't been tested well enough yet? If the former, what's remaining to do and would you accept PRs?

**Comment by matthiasplappert** [comment]
The main problem is that the asynchronous set up doesn't really work very well with OpenAI Gym (some environments crash when multiple instances are run simultaneously, if I recall correctly). Similarily, I also had problems with TensorFlow when attempting to use it from different threads. It has been a while since I have last worked on it and I *think* that what is currently on the `experimental` branch worked, but I am honestly not sure. However, if you want you can definitely try it out. Of course pull requests would be more than welcome and I am also happy to help out as much as I can to get it into a more stable form. Maybe I have some spare time next week to have another look at the current state.

**Comment by bear0330** [comment]
Looks like the problem is because OpenAI Gym? If I need to use A3C in my own problem not for Gym (via implement my own `Env` and `Space`), is the A3C implementation already done? 
By the way, do you have any plan to implement "Reinforcement learning with unsupervised auxiliary tasks." (https://arxiv.org/pdf/1611.05397.pdf) ? 
I found a very interesting implementation: https://github.com/miyosuda/unreal
But keras-rl is the library I wish to use, your work are so amazing.

Thanks.

**Comment by bear0330** [comment]
The problem @matthiasplappert mentioned about asynchronous issue, maybe you can reference this implementation https://github.com/NVlabs/GA3C, I didn't deeply look into it, but I know it implements A3C with GPU. If they can solve asynchronous problem, it might be a good reference. Thanks.

**Comment by bear0330** [comment]
@matthiasplappert : There is another implementation which uses A3C with Keras (not limited to Tensorflow), you might reference: https://github.com/lancelee82/rockrose
And it even implemented UNREAL algorithm with Keras.

**Comment by matthiasplappert** [comment]
Thanks for the pointers, I'll check that out!

**Comment by ahundt** [comment]
TensorPack also has a good implementation of a3c, though it is tf only:
https://github.com/ppwwyyxx/tensorpack/tree/master/examples/A3C-Gym

**Comment by raopku** [comment]
@matthiasplappert 
can you give some detailed information about the problem. i am developing A3C & Async DQN methods.
(1)The main problem is that the asynchronous set up doesn't really work very well with OpenAI Gym (some environments crash when multiple instances are run simultaneously, if I recall correctly). 

(2)Similarily, I also had problems with TensorFlow when attempting to use it from different threads.

I run the A3C method on the experiment branch, it works.

**Comment by wulabs** [comment]
@matthiasplappert With a few small patches to get code working against master, it does run, albeit is unstable and can crash in training.  In particular, due to theano not being thread safe (see below). Also, the initial model compile takes forever on my machine, about 30 min as compared to instanteous for the dqn and other examples.

https://github.com/Theano/Theano/issues/2186
`Training for 1000000 steps ...
Interval 1 (0 steps performed)
    5/10000 [..............................] - ETA: 74s - reward: -1.8774Assertion failed: (self->var_computed[input_idx]), function lazy_rec_eval, file /Users/*/.theano/compiledir_Darwin-16.3.0-x86_64-i386-64bit-i386-2.7.8-64/lazylinker_ext/mod.cpp, line 693.
Abort trap: 6`
